### ISSUE TITLE
Offer more (some!) approaches for template testing?

### ISSUE THREAD

**Issue by seancorfield** [seancorfield/deps-new#3]
Dominic Monroe on Slack:

> An approach to testing, even just a few basic utilities to spin up a repl of some kind and run some functions, testing that they don't explode would be great. Right now I have a bash script that generates a few variants of the project and then I manually check that they do sensible things (like that `(dev)` works, or that a server starts)

**Comment by SevereOverfl0w** [comment]
https://git.sr.ht/~severeoverfl0w/juxt-wedge/tree/master/item/test-template here's the script I use which generates multiple varieties.  There's a bunch of things I would ideally test as I change things (e.g. after running `(go)` does an HTTP request to / work?).

I can appreciate that some of these might be out of scope for clj-new, but I expect they're also the most useful thing to test alongside a template.  There're properties you want to have hold across a matrix of possible combinations of options.

**Comment by seancorfield** [comment]
Just so I understand, this is about some additional support for testing a _new template as you are developing it_ specifically, rather than general project testing?

**Comment by SevereOverfl0w** [comment]
Yeah, I'd be interested in testing the template in some way.  I suppose the tools I might need are:

* A tool to generate a full matrix of all possible options
* A way to run in the context of the generated projects, and run some kind of test suite in that context.  Perhaps forking out a new JVM for each project and running a script that I provide that could do the testing itself.

**Comment by SevereOverfl0w** [comment]
I ended up with fairly simple [tests](https://git.sr.ht/~severeoverfl0w/wedge-template/commit/208a84702c470ab279be64062974375593b53e86).

1. Start JVM with pREPL
2. Run a few commands in that pREPL & check they do what I expect
3. Test the system actually came up (in this case by using etaoin to load a browser and take a picture)
4. Shut it all down cleanly

It would be nice to encourage template developers to do 1/2/4 by having something for this in the default template template.  I rolled my own tools for that, I'm not sure if there's a programmatic client around for pREPL though.

**Comment by seancorfield** [comment]
I'm not convinced there's anything generic enough that the default `template` template could do to help here, since each project template is going to be (mostly) unique, with its own approach to how things work in the generated project, even in terms of just testing fns.

There is a test in the `template` template that the `template.edn` for the template is valid per Spec but that's it in terms of generic testing.

With the latest refactoring for `:test-runner` and `:build` and the new `transformers` ns, there are tests in `deps-new` itself that developers could use as a model for writing their own tests around `:data-fn` and `:template-fn`.

Closing this for now but open to PRs in the future for specific additions to the `template` template that may help with generic testing.
